File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/robotcrosswords/domain/RobotCrosswordsGame.java
package com.zwstudio.logicpuzzlesandroid.puzzles.robotcrosswords.domain;

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame;
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface;
import com.zwstudio.logicpuzzlesandroid.common.domain.Position;
import com.zwstudio.logicpuzzlesandroid.home.domain.HintState;

import java.util.ArrayList;
import java.util.List;

import fj.F2;
import fj.function.Effect1;

public class RobotCrosswordsGame extends CellsGame<RobotCrosswordsGame, RobotCrosswordsGameMove, RobotCrosswordsGameState> {
    public static Position offset[] = {
            new Position(-1, 0),
            new Position(0, 1),
            new Position(1, 0),
            new Position(0, -1),
    };

    public int[] objArray;
    public int get(int row, int col) {
        return objArray[row * cols() + col];
    }
    public int get(Position p) {
        return get(p.row, p.col);
    }
    public void set(int row, int col, int obj) {
        objArray[row * cols() + col] = obj;
    }
    public void set(Position p, int obj) {
        set(p.row, p.col, obj);
    }

    public List<List<Position>> areas = new ArrayList<>();
    public int horzAreaCount = 0;

    public RobotCrosswordsGame(List<String> layout, GameInterface<RobotCrosswordsGame, RobotCrosswordsGameMove, RobotCrosswordsGameState> gi, GameDocumentInterface gdi) {
        super(gi, gdi);
        size = new Position(layout.size(), layout.get(0).length());
        objArray = new int[rows() * cols()];

        for (int r = 0; r < rows(); r++) {
            String str = layout.get(r);
            for (int c = 0; c < cols(); c++) {
                char ch = str.charAt(c);
                set(r, c, ch == '.' ? -1 : ch == ' ' ? 0 : ch - '0');
            }
        }

        List<Position> area = new ArrayList<>();
        Effect1<Boolean> f = isHorz -> {
            if (area.isEmpty()) return;
            if (area.size() > 1) {
                areas.add(new ArrayList<>(area));
                if (isHorz) horzAreaCount++;
            }
            area.clear();
        };
        for (int r = 0; r < rows(); r++) {
            for (int c = 0; c < cols(); c++) {
                Position p = new Position(r, c);
                if (get(p) == -1)
                    f.f(true);
                else
                    area.add(p);
            }
            f.f(true);
        }
        for (int c = 0; c < cols(); c++) {
            for (int r = 0; r < rows(); r++) {
                Position p = new Position(r, c);
                if (get(p) == -1)
                    f.f(false);
                else
                    area.add(p);
            }
            f.f(false);
        }

        RobotCrosswordsGameState state = new RobotCrosswordsGameState(this);
        states.add(state);
        levelInitilized(state);
    }

    private boolean changeObject(RobotCrosswordsGameMove move, F2<RobotCrosswordsGameState, RobotCrosswordsGameMove, Boolean> f) {
        if (canRedo()) {
            states.subList(stateIndex + 1, states.size()).clear();
            moves.subList(stateIndex, states.size()).clear();
        }
        RobotCrosswordsGameState state = cloner.deepClone(state());
        boolean changed = f.f(state, move);
        if (changed) {
            states.add(state);
            stateIndex++;
            moves.add(move);
            moveAdded(move);
            levelUpdated(states.get(stateIndex - 1), state);
        }
        return changed;
   }

    public boolean switchObject(RobotCrosswordsGameMove move) {
        return changeObject(move, RobotCrosswordsGameState::switchObject);
    }

    public boolean setObject(RobotCrosswordsGameMove move) {
        return changeObject(move, RobotCrosswordsGameState::setObject);
    }

    public int getObject(Position p) {
        return state().get(p);
    }

    public int getObject(int row, int col) {
        return state().get(row, col);
    }

    public HintState getHorzState(Position p) {
        return state().pos2horzState.get(p);
    }

    public HintState getVertState(Position p) {
        return state().pos2vertState.get(p);
    }
}

File: app/src/main/java/com/zwstudio/logicpuzzlesandroid/puzzles/robotcrosswords/domain/RobotCrosswordsGame.kt
package com.zwstudio.logicpuzzlesandroid.puzzles.robotcrosswords.domain

import com.zwstudio.logicpuzzlesandroid.common.data.GameDocumentInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.CellsGame
import com.zwstudio.logicpuzzlesandroid.common.domain.GameInterface
import com.zwstudio.logicpuzzlesandroid.common.domain.Position
import java.util.*

class RobotCrosswordsGame(layout: List<String>, gi: GameInterface<RobotCrosswordsGame, RobotCrosswordsGameMove, RobotCrosswordsGameState>, gdi: GameDocumentInterface) : CellsGame<RobotCrosswordsGame, RobotCrosswordsGameMove, RobotCrosswordsGameState>(gi, gdi) {
    companion object {
        var offset = arrayOf(
            Position(-1, 0),
            Position(0, 1),
            Position(1, 0),
            Position(0, -1)
        )
    }

    var objArray: IntArray
    var areas = mutableListOf<List<Position>>()
    var horzAreaCount = 0

    operator fun get(row: Int, col: Int) = objArray[row * cols() + col]
    operator fun get(p: Position) = this[p.row, p.col]
    operator fun set(row: Int, col: Int, obj: Int) {objArray[row * cols() + col] = obj}
    operator fun set(p: Position, obj: Int) {this[p.row, p.col] = obj}

    init {
        size = Position(layout.size, layout[0].length)
        objArray = IntArray(rows() * cols())
        for (r in 0 until rows()) {
            val str = layout[r]
            for (c in 0 until cols()) {
                val ch = str[c]
                this[r, c] = if (ch == '.') -1 else if (ch == ' ') 0 else ch - '0'
            }
        }
        val area = mutableListOf<Position>()
        fun f(isHorz: Boolean) {
            if (area.isEmpty()) return
            if (area.size > 1) {
                areas.add(ArrayList(area))
                if (isHorz) horzAreaCount++
            }
            area.clear()
        }
        for (r in 0 until rows()) {
            for (c in 0 until cols()) {
                val p = Position(r, c)
                if (this[p] == -1)
                    f(true)
                else
                    area.add(p)
            }
            f(true)
        }
        for (c in 0 until cols()) {
            for (r in 0 until rows()) {
                val p = Position(r, c)
                if (this[p] == -1)
                    f(false)
                else
                    area.add(p)
            }
            f(false)
        }
        val state = RobotCrosswordsGameState(this)
        states.add(state)
        levelInitilized(state)
    }

    private fun changeObject(move: RobotCrosswordsGameMove, f: (RobotCrosswordsGameState, RobotCrosswordsGameMove) -> Boolean): Boolean {
        if (canRedo) {
            states.subList(stateIndex + 1, states.size).clear()
            moves.subList(stateIndex, states.size).clear()
        }
        val state = cloner.deepClone(state())
        val changed = f(state, move)
        if (changed) {
            states.add(state)
            stateIndex++
            moves.add(move)
            moveAdded(move)
            levelUpdated(states[stateIndex - 1], state)
        }
        return changed
    }

    fun switchObject(move: RobotCrosswordsGameMove) = changeObject(move, RobotCrosswordsGameState::switchObject)
    fun setObject(move: RobotCrosswordsGameMove) = changeObject(move, RobotCrosswordsGameState::setObject)

    fun getObject(p: Position) = state()[p]
    fun getObject(row: Int, col: Int) = state()[row, col]
    fun getHorzState(p: Position) = state().pos2horzState[p]
    fun getVertState(p: Position) = state().pos2vertState[p]
}
